ssl/statem/extensions.c

```diff
@@ -1731,3 +1747,116 @@ static int final_psk(SSL_CONNECTION *s, unsigned int context, int sent)
+                                                     X509 *x, size_t chainidx)
+{
+#ifndef OPENSSL_NO_COMP_ALG
+    int i;
+
+    if (!ossl_comp_has_alg(0))
+        return EXT_RETURN_NOT_SENT;
+
+    /* Do not indicate we support receiving compressed certificates */
+    if ((sc->options & SSL_OP_NO_RX_CERTIFICATE_COMPRESSION) != 0)
+        return EXT_RETURN_NOT_SENT;
+
+    if (sc->cert_comp_prefs[0] == TLSEXT_comp_cert_none)
+        return EXT_RETURN_NOT_SENT;
+
+    if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_compress_certificate)
+            || !WPACKET_start_sub_packet_u16(pkt)
+            || !WPACKET_start_sub_packet_u8(pkt))
+        goto err;
+
+    for (i = 0; sc->cert_comp_prefs[i] != TLSEXT_comp_cert_none; i++) {
+        if (!WPACKET_put_bytes_u16(pkt, sc->cert_comp_prefs[i]))
+            goto err;
+    }
+    if (!WPACKET_close(pkt) || !WPACKET_close(pkt))
+        goto err;
+
+    sc->ext.compress_certificate_sent = 1;
+    return EXT_RETURN_SENT;
+ err:
+    SSLfatal(sc, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+    return EXT_RETURN_FAIL;
+#else
+    return EXT_RETURN_NOT_SENT;
+#endif
+}
+
+#ifndef OPENSSL_NO_COMP_ALG
+static int tls_comp_idx(SSL_CONNECTION *sc, int alg)
+{
+    int i;
+
+    for (i = 0; i < TLSEXT_comp_cert_limit; i++)
+        if (sc->cert_comp_prefs[i] == alg)
+            return i;
+    /* if not found, return the limit */
+    return TLSEXT_comp_cert_limit;
+}
+#endif
+
+int tls_parse_compress_certificate(SSL_CONNECTION *sc, PACKET *pkt, unsigned int context,
+                                   X509 *x, size_t chainidx)
+{
+#ifndef OPENSSL_NO_COMP_ALG
+    PACKET supported_comp_algs;
+    unsigned int comp;
+    int idx;
+    int saved_idx;
+
+    if (!ossl_comp_has_alg(0))
+        return 1;
+
+    /* Ignore the extension and don't send compressed certificates */
+    if ((sc->options & SSL_OP_NO_TX_CERTIFICATE_COMPRESSION) != 0)
+        return 1;
+
+    if (!PACKET_as_length_prefixed_1(pkt, &supported_comp_algs)
+            || PACKET_remaining(&supported_comp_algs) == 0) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+        return 0;
+    }
+
+    if (sc->cert_comp_prefs[0] != TLSEXT_comp_cert_none) {
+        /*
+         * The preference array has real values, so take a look at each
+         * value coming in, and compare against our preference list.
+         * Pick the one closest to be beginning (i.e. lowest index)
+         * If a value is not in the list, it gets the limit value
+         * The array is 0 (i.e. "none") terminated
+         * The preference list only contains supported algorithms
+         */
+        saved_idx = tls_comp_idx(sc, sc->ext.compress_certificate_from_peer);
+        while (PACKET_get_1(&supported_comp_algs, &comp)) {
+            if ((idx = tls_comp_idx(sc, comp)) != TLSEXT_comp_cert_limit
+                    && idx < saved_idx) {
+                sc->ext.compress_certificate_from_peer = comp;
```

Comment:
If we are a server we should only select this if we already have the certificate precompressed.

Reply:
This could get complicated very quickly, as it's possible to have a sparsely defined set of compressed certificates.
e.g. 

- ECDSA compressed with brotli
- RSA compressed with zstd

At the parsing of the extensions, we don't know which certificate will be chosen (if I'm wrong, please let me know). So we accept what's possible. It's really at the point of writing out either a CERTIFICATE or COMPRESSED_CERTIFICATE the determination should be made. Perhaps this code just needs to save the array from the peer, and when we're ready to write out the certificate, that's when the determination of algorithm should be made. As it's the chosen certificate and the compressed data associated with it that really determines whether to compress or not, and how.